src/libraries/System.Private.CoreLib/src/System/Resources/ResourceReader.cs

```diff
@@ -459,68 +464,74 @@ private unsafe string AllocateStringForNameIndex(int index, out int dataOffset)
         }
 
         // This takes a virtual offset into the data section and reads a String
-        // from that location.
-        // Anyone who calls LoadObject should make sure they take a lock so
-        // no one can cause us to do a seek in here.
+        // from that location. Called from RuntimeResourceSet
         internal string? LoadString(int pos)
         {
             Debug.Assert(_store != null, "ResourceReader is closed!");
-            _store.BaseStream.Seek(_dataSectionOffset + pos, SeekOrigin.Begin);
-            string? s = null;
-            int typeIndex = _store.Read7BitEncodedInt();
-            if (_version == 1)
-            {
-                if (typeIndex == -1)
-                    return null;
-                if (FindType(typeIndex) != typeof(string))
-                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ResourceNotString_Type, FindType(typeIndex).FullName));
-                s = _store.ReadString();
-            }
-            else
+
+            lock (this)
```

Comment:
> it seems like the _store is only ever initialized with a Stream that's under our control, right?

@stephentoub I don't think that's correct. There is a public `ResourceReader(Stream)` constructor that populates `_store` with the user `Stream`. Of course most users are probably going through `ResourceManager` / `RuntimeResourceSet` where the runtime owns the stream. 

However, when it comes to calling arbitrary user code while holding the lock, this class is _already doing that_ because locking on `this` is `ResourceReader`'s primary existing synchronization mechanism. For example, consider the following code:
```
void Main()
{
	var myStream = new MyStream();
	var reader = new ResourceReader(myStream);
	myStream.Reader = reader;
	// prints "Calling user code from within locked ResourceReader" many times
	foreach (DictionaryEntry entry in reader) { }
}

class MyStream : MemoryStream
{
	public ResourceReader Reader { get; set; }
	
	public MyStream() : base(GetBytes()) { }
	
	private static byte[] GetBytes()
	{
		using var stream = typeof(Expression).Assembly.GetManifestResourceStream("FxResources.System.Linq.Expressions.SR.resources");
		var bytes = new byte[stream.Length];
		if (stream.Read(bytes, 0, bytes.Length) != bytes.Length) { throw new Exception(); }
		return bytes;
	}
	
	public override int Read(byte[] buffer, int offset, int count)
	{
		if (this.Reader != null && Monitor.IsEntered(this.Reader)) 
		{ 
			Console.WriteLine("Calling user code from within locked ResourceReader");
		}
		return base.Read(buffer, offset, count);
	}
}
```

Reply:
Any thoughts with this @stephentoub ? My view is that since we were already using this locking approach merely extending it to be used consistently shouldn't be a concern. 

Today, `ResourceReader` is "mostly" thread-safe apart from a couple methods missing locks. If we wanted to avoid this, I think we'd need to remove all locking from `ResourceReader` and lock in `RuntimeResourceSet` instead to preserve thread-safety. This could be a breaking change for people relying on `ResourceReader`'s thread-safety, though.